### PR TITLE
o/snapstate: install components from the store

### PR DESCRIPTION
**Pull request by andrewphelpsj**
This PR enables us to install components and a snap from the store at the same time. Note that installing components from the store for an already installed snap is not yet handled yet, as that interaction with the store is more complex, requiring the usage of both channel and revision.

First relevant commit is 97dad459a4e9410994c02942c795b882c3e52791, based on https://github.com/snapcore/snapd/pull/14070.

### CODE REVIEW

File: overlord/snapstate/target.go [hunk 1]
```diff
@@ -266,37 +294,92 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+	compName := naming.NewComponentRef(info.SnapName(), name)
+
+	return componentTarget{
+		setup: ComponentSetup{
+			DownloadInfo: &sar.DownloadInfo,
+		},
+		info: &snap.ComponentInfo{
+			Component: compName,
+			Type:      comp.Type,
+			Version:   sar.Version,
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/target.go:366]: Should we sanity chech the version?

**Review by alfonsosanchezbeato — COMMENTED** [review]
Some comments, see below

File: overlord/snapstate/target.go [hunk 2]
```diff
@@ -110,7 +116,25 @@ func (t *target) snapsup(st *state.State, opts Options) (SnapSetup, error) {
...
+// componentTarget represents the data needed to setup a component for installation.
+type componentTarget struct {
+	// setup is a partially initialized ComponentSetup struct that contains the
+	// data needed to find the component that will be installed.
+	setup ComponentSetup
+	// info contains the snap.ComponentInfo for the component to be installed.
+	info *snap.ComponentInfo
+}
+
+func (c *componentTarget) compsup() ComponentSetup {
```
[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:131]: It is shared here since there will be another implementation of `UpdateGoal` which will also return a target.

**Comment by andrewphelpsj** [comment]
@alfonsosanchezbeato as for the spread test, not yet without implementing some of the CLI aspects changes. Would you like that as a part of this PR?

**Comment by alfonsosanchezbeato** [comment]
> @alfonsosanchezbeato as for the spread test, not yet without implementing some of the CLI aspects changes. Would you like that as a part of this PR?

Nope, we can do it in a follow-up, maybe as you send when we have api/cli changes.

File: overlord/snapstate/target.go (hunk 1 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:366]: `snap.Component` doesn't carry a `Version`, is there something else you were thinking to compare it to?

Comment (alfonsosanchezbeato) [reply overlord/snapstate/target.go:366]: It would need to be read from component.yaml, I guess we can skip. Not sure though if what we put in `store.SnapActionResult` is checked against snap.yaml or not.

File: overlord/snapstate/target.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (alfonsosanchezbeato) [reply overlord/snapstate/target.go:131]: But I do not see any mention to `UpdateGoal` in this PR (or am I missing something?), maybe it will become clear in the next PR, but atm I find this a bit confusing, maybe this part of the change could go in next one?

[the remark this answers is not in the document]
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:131]: Sorry, I meant `InstallGoal`. Too many PRs in my head :slightly_smiling_face:. I can move this to the next PR, which includes an `InstallGoal` implementation that installs from local files.

File: overlord/snapstate/component.go [hunk 3]
```diff
@@ -71,15 +71,21 @@ func InstallComponentPath(st *state.State, csi *snap.ComponentSideInfo, info *sn
...
+
+	return doInstallComponent(st, &snapst, compSetup, snapsup, componentInstallFlags{
+		// The file passed around is temporary, make sure it gets removed.
+		RemoveComponentPath: true,
+	}, "")
+}
+
+type componentInstallFlags struct {
+	RemoveComponentPath bool
+	SkipSecurity        bool
```
Comment (pedronis) [inline overlord/snapstate/component.go:83]: this is better named SkipProfiles

File: overlord/snapstate/handlers_components.go [hunk 4]
```diff
@@ -222,6 +295,16 @@ func (m *SnapManager) doMountComponent(t *state.Task, _ *tomb.Tomb) (err error)
 	perfTimings.Save(st)
 	st.Unlock()
 
+	// TODO: test this conditional once we are installing local snaps+components
+	// if we're removing the snap file and we are mounting a component for the
+	// firs time, then we know that the component also must be coming from an
+	// emphemeral file. in that case, remove it.
+	if snapsup.RemoveSnapPath {
+		if err := os.Remove(compSetup.CompPath); err != nil {
```
Comment (pedronis) [inline overlord/snapstate/handlers_components.go:303]: this seems not be tested

File: overlord/snapstate/snapstate_remove_test.go [hunk 5]
```diff
@@ -438,7 +438,7 @@ func (s *snapmgrTestSuite) TestParallelInstanceRemoveRunThrough(c *C) {
 		{
 			op:             "remove-snap-data-dir",
 			name:           "some-snap_instance",
-			path:           filepath.Join(dirs.SnapDataDir, "some-snap"),
+			path:           filepath.Join(dirs.SnapDataDir, "some-snap_instance"),
```
Comment (pedronis) [inline overlord/snapstate/snapstate_remove_test.go:441]: why this change?

File: overlord/snapstate/target.go [hunk 6]
```diff
@@ -110,7 +116,16 @@ func (t *target) snapsup(st *state.State, opts Options) (SnapSetup, error) {
 		PlugsOnly:          len(t.info.Slots) == 0,
 		InstanceKey:        t.info.InstanceKey,
 		ExpectedProvenance: t.info.SnapProvenance,
-	}, nil
+	}, compsups, nil
+}
+
+// componentTarget represents the data needed to setup a component for installation.
+type componentTarget struct {
```
Comment (pedronis) [inline overlord/snapstate/target.go:123]: this should be closer to the other struct

File: overlord/snapstate/target.go [hunk 7]
```diff
@@ -266,37 +285,96 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+}
+
 func installActionForStoreTarget(t StoreSnap, opts Options, enforcedSets func() (*snapasserts.ValidationSets, error)) (*store.SnapAction, error) {
 	action := &store.SnapAction{
 		Action:       "install",
 		InstanceName: t.InstanceName,
 		Channel:      t.RevOpts.Channel,
 		Revision:     t.RevOpts.Revision,
 		CohortKey:    t.RevOpts.CohortKey,
+		Resources:    t.Components,
```
Comment (pedronis) [inline overlord/snapstate/target.go:377]: what's the reason we need to pass this along? for metrics?

File: overlord/snapstate/target.go [hunk 8]
```diff
@@ -110,7 +116,16 @@ func (t *target) snapsup(st *state.State, opts Options) (SnapSetup, error) {
...
+	}, compsups, nil
+}
+
+// componentTarget represents the data needed to setup a component for installation.
+type componentTarget struct {
+	// setup is a partially initialized ComponentSetup struct that contains the
+	// data needed to find the component that will be installed.
+	setup ComponentSetup
+	// info contains the snap.ComponentInfo for the component to be installed.
+	info *snap.ComponentInfo
```
Comment (pedronis) [inline overlord/snapstate/target.go:128]: where is this field used?

**Review by pedronis — COMMENTED** [review]
did a pass, some comments/questions, some might be for the store

File: overlord/snapstate/backend_test.go [hunk 9]
```diff
@@ -1372,7 +1405,7 @@ func (f *fakeSnappyBackend) RemoveSnapDir(s snap.PlaceInfo, otherInstances bool)
 	f.appendOp(&fakeOp{
 		op:             "remove-snap-dir",
 		name:           s.InstanceName(),
-		path:           snap.BaseDir(s.SnapName()),
+		path:           snap.BaseDir(s.InstanceName()),
```
Comment (andrewphelpsj) [inline overlord/snapstate/backend_test.go:1408]: This change here I think is better representative of what actually happens in `RemoveSnapDataDir` (the real implementation).

File: overlord/snapstate/snapstate_remove_test.go (hunk 5 above)
Comment (andrewphelpsj) [reply overlord/snapstate/snapstate_remove_test.go:441]: It lines up with this change here: https://github.com/snapcore/snapd/pull/14092/files#r1647912200.

It doesn't have to happen, but I noticed that it was inconsistent when writing the new tests here.

File: overlord/snapstate/target.go (hunk 7 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:377]: It is ultimately used to indicate in the store package that we should request resource info from the store. It could (and maybe should?) be a boolean value. It is a bit weird,  since the option is a per-request toggle (just a query param), but right here I have it expressed as a per-action concept. But the `StoreService.SnapAction` API doesn't have any sort of flags right now.

File: overlord/snapstate/target.go (hunk 8 above)
Comment (andrewphelpsj) [reply overlord/snapstate/target.go:128]: It can be eliminated in the current version of the code. I originally had it there because of how the `InstallGoal` implementation filled in the `ComponentSetup` struct. It partially filled it in, but that is no longer true. In that case, the entire `componentTarget` struct can go away.

File: overlord/snapstate/snapstate.go [hunk 10]
```diff
@@ -626,7 +668,10 @@ func doInstall(st *state.State, snapst *SnapState, snapsup *SnapSetup, flags int
 	// additional tasks.
 	linkSnap := st.NewTask("link-snap", fmt.Sprintf(i18n.G("Make snap %q%s available to the system"), snapsup.InstanceName(), revisionStr))
 	addTask(linkSnap)
-	prev = linkSnap
+
+	for _, t := range tasksAfterLinkSnap {
+		addTask(t)
```
Comment (pedronis) [inline overlord/snapstate/snapstate.go:673]: does this match where we specced after link hooks for components should run vs snap own hooks?

**Review by pedronis — COMMENTED** [review]
thanks for changes, one question

File: overlord/snapstate/snapstate.go (hunk 10 above)
Comment (andrewphelpsj) [reply overlord/snapstate/snapstate.go:673]: The component hook tasks aren't created yet. Should I add a TODO here for once https://github.com/snapcore/snapd/pull/13775 gets merged?

File: overlord/snapstate/handlers_components.go (hunk 4 above)
Comment (andrewphelpsj) [reply overlord/snapstate/handlers_components.go:303]: This change shouldn't have been in this PR, it should be in the next one. My mistake.

File: overlord/snapstate/target.go [hunk 11]
```diff
@@ -266,30 +278,80 @@ func (s *storeInstallGoal) toInstall(ctx context.Context, st *state.State, opts
...
+	comp, ok := info.Components[name]
+	if !ok {
+		return ComponentSetup{}, fmt.Errorf("%q is not a component for snap %q", name, info.SnapName())
+	}
+
+	if typ := fmt.Sprintf("component/%s", comp.Type); typ != sar.Type {
+		return ComponentSetup{}, fmt.Errorf("inconsistent component type (%q in snap, %q in component)", typ, sar.Type)
+	}
+
+	compName := naming.NewComponentRef(info.SnapName(), name)
```
Comment (alfonsosanchezbeato) [inline overlord/snapstate/target.go:341]: nitpick: maybe cref, there are too many `*name` variables here

File: overlord/snapstate/snapstate.go (hunk 10 above)
Comment (alfonsosanchezbeato) [reply overlord/snapstate/snapstate.go:677]: As discussed, we are not adding hook tasks yet in `doInstallComponent()`, but that will eventually happen. In any case, if `componentTasksForInstallWithSnap()` keeps the tasks order things should be fine, maybe for the moment we could just add a comment there commenting that order needs to be kept.

**Review by alfonsosanchezbeato — APPROVED** [review]
LGTM, couple of minor comments

**Review by pedronis — APPROVED** [review]
thanks
